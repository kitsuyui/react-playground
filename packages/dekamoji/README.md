# @kitsuyui/react-dekamoji

[![npm version](https://badge.fury.io/js/@kitsuyui%2Freact-dekamoji.svg)](https://badge.fury.io/js/@kitsuyui%2Freact-dekamoji)

Dekamoji is a React component that changes the font size according to the size of the parent element. It displays the largest possible font size that can be displayed in response to changes in the size of the parent element.

## Demo

Storybook: https://react-playground.docs.kitsuyui.com/storybook/

## Installation

### npm

```sh
npm install @kitsuyui/react-dekamoji
```

### yarn

```sh
yarn add @kitsuyui/react-dekamoji
```

### pnpm

```sh
pnpm add @kitsuyui/react-dekamoji
```

## License

MIT
