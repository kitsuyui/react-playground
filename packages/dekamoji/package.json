{
  "name": "@kitsuyui/react-dekamoji",
  "version": "0.0.0",
  "description": "A component that displays text in the entire display area of the component.",
  "repository": {
    "type": "git",
    "url": "https://github.com/kitsuyui/react-playground.git"
  },
  "license": "MIT",
  "author": "Yui Kitsu <kitsuyui@kitsuyui.com>",
  "type": "module",
  "exports": {
    ".": {
      "require": {
        "type": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      },
      "import": {
        "type": "./dist/index.d.ts",
        "default": "./dist/index.js"
      }
    }
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "package.json"
  ],
  "scripts": {
    "build": "tsup --clean",
    "dev": "tsup --watch",
    "test": "vitest run",
    "validate": "validate-package-exports --check --verify"
  },
  "devDependencies": {
    "@types/node": "^24.6.2",
    "@types/react": "^19.1.16",
    "@types/react-dom": "^19.1.9",
    "react": "^19.1.1",
    "react-dom": "^19.1.1",
    "react-use": "^17.6.0"
  },
  "peerDependencies": {
    "react": "*",
    "react-dom": "*"
  }
}
