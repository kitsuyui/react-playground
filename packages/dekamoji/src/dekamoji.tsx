import React, { useRef, useMemo } from 'react'
import { useMeasure } from 'react-use'

type Props = {
  text: string
}

export const Dekamoji: React.FC<Props> = React.memo(function Dekamoji({
  text,
}: {
  text: string
}): React.JSX.Element {
  const [ref, { width, height }] = useMeasure<HTMLDivElement>()
  const [fontSize, setFontSize] = React.useState(0)
  const ref2 = useRef<HTMLDivElement>(null)

  useMemo(() => {
    const element = ref2.current
    let fontFamily: string | undefined
    if (element) {
      fontFamily = detectFontFamily(element)
    }
    const size = calcFontSize(width, height, text, fontFamily)
    setFontSize(size)
  }, [width, height, text])

  return (
    <div
      ref={ref}
      style={{
        width: '100%',
        height: '100%',
        boxSizing: 'border-box',
        overflow: 'hidden',
      }}
    >
      <div
        ref={ref2}
        style={{
          position: 'absolute',
          width: '100%',
          height: '100%',
          fontSize: `${fontSize}px`,
          textAlign: 'center',
          margin: '0 auto',
          boxSizing: 'border-box',
          whiteSpace: 'pre-wrap',
        }}
      >
        {text}
      </div>
    </div>
  )
})

/**
 * Detect font family inherited from parent element
 * @param element
 * @returns font family
 */
const detectFontFamily = (element: HTMLElement): string | undefined => {
  const style = window.getComputedStyle(element)
  const fontFamily = style.fontFamily
  return fontFamily
}

const calcFontSize = (
  width: number,
  height: number,
  text: string,
  fontFamily?: string
): number => {
  // calculate font size without react component and raw dom
  const outer = document.createElement('div')
  outer.style.position = 'absolute'
  outer.style.width = `${width}px`
  outer.style.height = `${height}px`
  outer.style.boxSizing = 'border-box'
  outer.style.zIndex = '-1'
  outer.style.overflowX = 'hidden'
  outer.style.overflowY = 'hidden'
  const inner = document.createElement('div')
  inner.style.visibility = 'hidden'
  inner.style.fontSize = '0'
  inner.style.textAlign = 'center'
  inner.style.margin = '0 auto'
  inner.style.whiteSpace = 'pre-wrap'
  inner.style.boxSizing = 'border-box'
  inner.style.zIndex = '-1'
  if (fontFamily) {
    inner.style.fontFamily = fontFamily
  }
  inner.textContent = text
  outer.appendChild(inner)
  document.body.appendChild(outer)

  // binary search for font size
  const maxFontSize = Math.max(width, height)
  const min = 0.0
  const max = maxFontSize
  let minOverflow = max
  let maxAvailable = min
  let trial = Math.floor((minOverflow + maxAvailable) / 2)
  const maxIterations = Math.log2(maxFontSize)
  let iterations = 0
  while (iterations < maxIterations) {
    iterations += 1
    trial = Math.floor((minOverflow + maxAvailable) / 2)
    inner.style.fontSize = `${trial}px`
    const overflowHeight = inner.scrollHeight - height
    const overflowWidth = inner.scrollWidth - width
    const scrollbarWidth = 2
    const overflows = overflowHeight > scrollbarWidth || overflowWidth > scrollbarWidth
    if (overflows) {
      minOverflow = trial
    } else {
      maxAvailable = trial
    }
    if (minOverflow - maxAvailable <= 1) {
      break
    }
  }
  document.body.removeChild(outer)
  return trial - 4
}
