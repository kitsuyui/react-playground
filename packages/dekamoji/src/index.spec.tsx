import { expect, test } from 'vitest'

import { render, screen } from '@testing-library/react'

import { Dekamoji } from './dekamoji'

test('render Dekamoji', () => {
  render(<Dekamoji text="あ" />)
  const element = screen.getAllByText(/あ/)
  expect(element).toHaveLength(1)
  expect(element).toMatchSnapshot()
})
