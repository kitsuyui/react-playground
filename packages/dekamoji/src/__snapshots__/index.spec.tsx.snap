// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render Dekamoji 1`] = `
[
  <div
    style="position: absolute; width: 100%; height: 100%; font-size: -4px; text-align: center; margin: 0px auto; box-sizing: border-box; white-space: pre-wrap;"
  >
    „ÅÇ
  </div>,
]
`;
